### PR TITLE
Make spawn usage consistent with a single way of calling.  

### PR DESCRIPTION
**Pull request by technorama**
Remove redundant macro.

### CODE REVIEW

**Comment by asterite** [comment]
It's not redundant. You forgot to change it in `samples/channel_primes.cr`. Change it to use blocks and you'll see the problem. The spawn macro that accepts a call will create a closure that copies the call arguments so you don't have the problem of accidentally referencing old values. That's why in go you can do: `go function(x, y)`, because it's basically the same as our second spawn macro.

**Comment by technorama** [comment]
I find `spawn(exp)` completely nonintuitive.

I read
  `spawn(handle_client(server.accept))` 
as
  `spawn { handle_client(server.accept) }`
or
  `Fiber.new(f2(f1()))` # with scheduling of course

That's not what's happening at all.

Instead `server.accept`/`f1()` is called OUTSIDE of the spawn block.  That completely breaks my concept of function call precedence.

I consider this behavior counterintuitive and a bug.

**Comment by waj** [comment]
I'll give you my viewpoint that I think makes the usage more clear. Compare these two examples:

```crystal
some_method(arg)
```
```crystal
spawn some_method(arg)
```

In the second case, I expect that the call performed is exactly the same, but asynchronous. That's all, I don't think in closures, just the method called and the argument given in the moment of execution.

In other words, I wouldn't try to think `spawn` as a regular method call. Instead I'd like to say it's a keyword that makes the program to continue running without waiting for the method call or the block to finish the execution.

**Comment by jhass** [comment]
I have to agree that it is a bit counter intuitive that the argument is not evaluated asynchronously but the call is, it isn't obvious. I'd say let's play a bit with alternative syntaxes

```cr
spawn handle_client, server.accept
spawn :handle_client, server.accept
spawn :handle_client => server.accept
spawn handle_client, when: server.accept
spawn handle_client, with: server.accept
spawn handle_client, after: server.accept
spawn handle_client, for: server.accept
spawn server.accept, handle_client
spawn server.accept, :handle_client
spawn server.accept, then: handle_client
spawn_for server.accept, handle_client
spawn(server.accept) {|socket| handle_client(socket) } # I think it's already possible for lvar's?
spawn handle_client << server.accept
with server.accept spawn handle_client
```
That would be my brainstorming for now. Any other ideas? Preferences?

**Comment by technorama** [comment]
Slightly related is a futures syntax from Celluloid.  
 `obj.future.method(arg)` returns a future.  I'd like to see this augment or replace future {}
Perhaps
`obj.async.method(arg)` used in Celluloid
or
`obj.spawn.method(arg)` 

If this can be done with macros that would be wonderful.  Actors will probably need to modify the behavior of `future` or `sync/spawn` calls slightly.

**Comment by ysbaddaden** [comment]
I don't see a change from:

```crystal
spawn handle_client(server.accept)
```

to anything else proposed here as being an improvement. It's either more code, or harder to grasp, especially with more args.

**Comment by technorama** [comment]
There's also the current inability to pass arguments to spawn which started my investigation of how spawn works.  Some method of passing arguments will be necessary for futures or async behavior.

```
async(executor: client_thread_pool).handle_client(server.accept)
```

**Comment by waj** [comment]
I still don't understand why you expect the arguments to be evaluated lazily. It's the call what should be seen as asynchronously executed, not the arguments.

BTW, we didn't invented this, we blamelessly copied that behaviour from Go, after finding the issue of closured variables when trying to convert some synchronous code to use spawn, and finding ourselves in what we initially thought more intuitive.

I know... you can argue that maybe Go didn't got this right also. But I find much more intuitive that, if I have a code like this:

```crystal
loop do
  s = server.accept
  handle_client(s)
end
```

and I wish to modify that program to start handling clients asynchronously, all I had to do is add the `spawn` before the call.

I think all these things cannot be 100% intuitive anyway, and most probably `spawn` and his friends require an entire dedicated section in the documentation.

File: samples/tcp_server.cr [hunk 1]
```diff
@@ -20,4 +20,7 @@ end
 
 server = TCPServer.new "127.0.0.1", 9000
 puts "listen on 127.0.0.1:9000"
-loop { spawn process(server.accept) }
+loop do
+  client = server.accept
+  spawn { process(client) }
```
Comment (waj) [inline samples/tcp_server.cr]: Notice here you are spawning with a closured variable, so you could end up processing the same client twice and loose others if the `server.accept` is executed again before the `process` method starts.

As you can see, you committed the same mistake we did on the beginning, so seems like this approach is not totally intuitive after all :smiley:

Comment (technorama) [inline samples/tcp_server.cr]: So neither behavior is intuitive.

Comment (waj) [inline samples/tcp_server.cr]: Exactly, that's what I said in my previous comment, but I think it's acceptable that there might be common pitfalls until someone learns how things work.

**Comment by technorama** [comment]
Go got a lot of things wrong.  They solve problems with lots of low level code.  I prefer more expressive code.

`future foo(bar())`
* Doesn't allow arguments
* Isn't intuitive.  Which parts of this run inside the future and which outside?
* Looks too similar to `future(foo(()))` which using method evaluation rules would imply the result of `foo(bar())` is passed to `future`.

`future.foo(bar())`
* Does allow arguments.
* Has a reasonable explanation.  This creates a new future and sends a message to it.
* Normal function evaluation rules apply.  `bar()` is always evaluated first, then `future` and finally `foo`.

**Pull request closed without merging**